### PR TITLE
Coerce values obtained from polyglot calls

### LINKED ISSUE #5177: Equality with polyglot numbers is broken
This task is automatically imported from the old Task Issue Board and it was originally created by Hubert Plociniczak.
Original issue is [here](https://www.pivotaltracker.com/story/show/183764270).
<hr />

```
int = Polyglot.new Integer [42]
int . should_equal 42
```
is currently false.
So is
```
42 . should_equal int
```
The main issue is that when the equality check sees that we compare a polyglot and a non-polyglot value it assumes that can't be ever true and bails.
We need to special case numbers.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Coerce values obtained from polyglot calls to fix #5177.

### Important Notes

Adds `IntHolder` class into the `test/Tests` project to simulate access to a class with integer field. 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: test/Tests/src/Data/Polyglot_Spec.enso [hunk 1]
```diff
@@ -19,9 +21,26 @@ spec = Test.group "Polyglot" <|
...
     Test.specify "should be able to execute a polyglot function object along with corresponding arguments" <|
         fun = Function.identity
         Polyglot.execute fun ["42"] . should_equal "42"
         Polyglot.execute fun ["42"].to_array . should_equal "42"
 
 main = Test_Suite.run_main spec
+
+
+foreign js js_meaning = """
+    return { meaning : 6 * 7 };
```
Comment (jdunkerley) [inline test/Tests/src/Data/Polyglot_Spec.enso]: can we keep "main" at the bottom please

File: test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/IntHolder.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.enso.base_test_helpers;
+
+public final class IntHolder {
+  public final Integer value;
```
Comment (jdunkerley) [inline test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/IntHolder.java]: Out of interest why box from `int` to `Integer`?

**Review by jdunkerley — APPROVED** [review]

File: test/Tests/src/Data/Polyglot_Spec.enso [hunk 3]
```diff
@@ -19,9 +21,26 @@ spec = Test.group "Polyglot" <|
         Polyglot.new String ["42"] . should_equal "42"
         Polyglot.new String ["42"].to_array . should_equal "42"
 
+    Test.specify "use Integer created by constructor" <|
+        Polyglot.new Integer [42] . should_equal 42
+
+    Test.specify "use Integer read from Polyglot object" <|
+        (Polyglot.get_member js_meaning "meaning") . should_equal 42
+
```
Comment (Akirathan) [inline test/Tests/src/Data/Polyglot_Spec.enso:33]:
```suggestion
        js_meaning.meaning . should_equal 42

```

**Review by Akirathan — APPROVED** [review]

File: test/Tests/src/Data/Polyglot_Spec.enso [hunk 4]
```diff
@@ -19,9 +21,26 @@ spec = Test.group "Polyglot" <|
         Polyglot.new String ["42"] . should_equal "42"
         Polyglot.new String ["42"].to_array . should_equal "42"
 
+    Test.specify "use Integer created by constructor" <|
+        Polyglot.new Integer [42] . should_equal 42
+
```
Comment (Akirathan) [inline test/Tests/src/Data/Polyglot_Spec.enso:27]:
```suggestion
        Polyglot.new Double [42.0] . should_equal 42

```

File: test/Tests/src/Data/Polyglot_Spec.enso (hunk 1 above)
Comment (JaroslavTulach) [reply test/Tests/src/Data/Polyglot_Spec.enso]: The textmate grammar that we currently use has problems with `"""` - then everything below is treated as a string including the `main` function. That's why I have a habit to put `foreign` functions to the bottom.

However you are right, we should fix the grammar.

Comment (JaroslavTulach) [reply test/Tests/src/Data/Polyglot_Spec.enso]: Fixed in f493c6d3f

File: test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/IntHolder.java (hunk 2 above)
Comment (JaroslavTulach) [reply test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/IntHolder.java]: Good question. Let's test both variants: a311e6345

File: test/Tests/src/Data/Polyglot_Spec.enso (hunk 3 above)
Comment (JaroslavTulach) [reply test/Tests/src/Data/Polyglot_Spec.enso:33]: The goal of the code is to test `Polyglot.get_member` and make sure it properly coerces its value to `long`. Testing field read as well: 158309dc1

File: test/Tests/src/Data/Polyglot_Spec.enso (hunk 4 above)
Comment (JaroslavTulach) [reply test/Tests/src/Data/Polyglot_Spec.enso:27]: Ok, let's test `Double` as well: 0262db733

File: test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/IntHolder.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+public final class IntHolder {
+  public final int value;
+  public final Integer boxed;
+
+  public IntHolder(int value) {
+    this.value = value;
+    this.boxed = value;
+  }
+}
```
Comment (radeusgd) [inline test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/IntHolder.java:11]: I'm glad that the test helpers project is getting use 🙂

**Review by 4e6 — APPROVED** [review]

**Review by radeusgd — APPROVED** [review]

**Review by GregoryTravis — APPROVED** [review]

**Pull request merged**